### PR TITLE
AutoGluon AutoTabPFN

### PR DESCRIPTION
**Pull request by klemens-floege**
Built Scikit-Learn Wrapper for AutoGluon with TabPFN model, inside old AutoTabPFN wrapper.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -59,7 +60,11 @@ interpretability = [
 ]
 post_hoc_ensembles = [
     "llvmlite",
-    "hyperopt>=0.2.7"
+    "hyperopt>=0.2.7",
+    "autogluon.common @ git+https://github.com/autogluon/autogluon#subdirectory=common",
+    "autogluon.features @ git+https://github.com/autogluon/autogluon#subdirectory=features",
+    "autogluon.core @ git+https://github.com/autogluon/autogluon#subdirectory=core",
+    "autogluon.tabular @ git+https://github.com/autogluon/autogluon#subdirectory=tabular",
```
[the remark this answers is not in the document]
Comment (klemens-floege) [reply pyproject.toml]: Not having a commit hash is advantageous to benefit from recent developments on Autogluon Tabular, ideally one the new package version is released we do not have to import from GitHub anymore anyway.

[the remark this answers is not in the document]
Comment (klemens-floege) [reply pyproject.toml]: See above

File: src/tabpfn_extensions/post_hoc_ensembles/utils.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+def search_space_func(
+    task_type: Literal["regression", "multiclass"],
+    num_random_configs: int = 200,
+    seed=1234,
+) -> list[dict]:
+    """Generate a list of random configurations for TabPFNv2 from Search Space,
+    which can be used as by Autogluon to construct a ensemble model.
+
+    Also includes lazy imports for hyperopt since it is not in base requirements of package
+    """
```
[the remark this answers is not in the document]
Comment (klemens-floege) [reply src/tabpfn_extensions/post_hoc_ensembles/utils.py]: done

File: src/tabpfn_extensions/post_hoc_ensembles/utils.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    search_space = get_param_grid_hyperopt(task_type=task_type)
+    rng = np.random.default_rng(seed)
+    stochastic.sample(search_space, rng=rng)
```
[the remark this answers is not in the document]
Comment (klemens-floege) [reply src/tabpfn_extensions/post_hoc_ensembles/utils.py]: Done

File: pyproject.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (klemens-floege) [reply pyproject.toml]: Will add, at least then nothing will break unexpectedly.

File: src/tabpfn_extensions/post_hoc_ensembles/utils.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    raw_config.pop("max_depth", None)
+
+    return raw_config
+
+
+def search_space_func(
+    task_type: Literal["regression", "multiclass"],
+    num_random_configs: int = 200,
+    seed=1234,
+) -> list[dict]:
```
[the remark this answers is not in the document]
Comment (klemens-floege) [reply src/tabpfn_extensions/post_hoc_ensembles/utils.py]: Done

File: src/tabpfn_extensions/post_hoc_ensembles/sklearn_interface.py [hunk 5]
```diff
@@ -145,35 +270,12 @@
...
-        self.predictor_.fit(
-            X,
-            y,
-            categorical_feature_indices=self.categorical_feature_indices,
-        )
-
-        # -- Sklearn required values
-        self.classes_ = self.predictor_._label_encoder.classes_
-        self.n_features_in_ = self.predictor_.n_features_in_
+        TaskType.MULTICLASS if len(self.classes_) > 2 else TaskType.BINARY
```
[the remark this answers is not in the document]
Comment (klemens-floege) [reply src/tabpfn_extensions/post_hoc_ensembles/sklearn_interface.py]: Done

File: src/tabpfn_extensions/post_hoc_ensembles/sklearn_interface.py [hunk 6]
```diff
@@ -45,76 +61,188 @@
...
+            The predictor interface used to make predictions.
         phe_init_args_ : dict
             The optional initialization arguments used for the post hoc ensemble predictor.
     """
 
-    predictor_: AutoPostHocEnsemblePredictor
-    phe_init_args_: dict
-    n_features_in_: int
-
     def __init__(
```
[the remark this answers is not in the document]
Comment (klemens-floege) [reply src/tabpfn_extensions/post_hoc_ensembles/sklearn_interface.py:105]: Done

File: src/tabpfn_extensions/post_hoc_ensembles/sklearn_interface.py [hunk 7]
```diff
@@ -1,141 +1,382 @@
...
+            num_gpus = torch.cuda.device_count()
+
+        self.predictor_.fit(
+            train_data=training_df,
+            time_limit=self.max_time,
+            presets=self.presets,
+            hyperparameters=hyperparameters,
+            num_gpus=num_gpus,
+            **self._get_predictor_fit_args(),
+        )
```
[the remark this answers is not in the document]
Comment (klemens-floege) [reply src/tabpfn_extensions/post_hoc_ensembles/sklearn_interface.py]: Done, however TabularPredictor.fit() does not accept a seed argument

File: pyproject.toml [hunk 8]
```diff
@@ -59,7 +60,11 @@ interpretability = [
 ]
 post_hoc_ensembles = [
     "llvmlite",
-    "hyperopt>=0.2.7"
+    "hyperopt>=0.2.7",
+    "autogluon.common @ git+https://github.com/autogluon/autogluon@7f12d1660ae4061f86a705f8c9b6b2ba83d310d5#subdirectory=common",
```
Comment (noahho) [inline pyproject.toml:66]:
```suggestion
    # Pinned to a specific autogluon commit to get access to AutoGluon TabPFN.
    # TODO: Move to a released version once autogluon 1.4 is available.
    "autogluon.common @ git+https://github.com/autogluon/autogluon@7f12d1660ae4061f86a705f8c9b6b2ba83d310d5#subdirectory=common",
```

File: tests/test_post_hoc_ensembles.py [hunk 9]
```diff
@@ -24,31 +51,40 @@ class TestAutoTabPFNClassifier(BaseClassifierTests):
     def estimator(self, tabpfn_classifier):
         """Provide a PHE-based TabPFN classifier as the estimator."""
         # For PHE, we can make tests faster by limiting time and using minimal models
-        max_time = 1 if FAST_TEST_MODE else 5  # Very limited time for fast testing
+        max_time = 15 if FAST_TEST_MODE else 30  # Very limited time for fast testing
```
Comment (noahho) [inline tests/test_post_hoc_ensembles.py]: why increased?

File: tests/test_post_hoc_ensembles.py [hunk 10]
```diff
@@ -60,32 +96,40 @@ class TestAutoTabPFNRegressor(BaseRegressorTests):
     def estimator(self, tabpfn_regressor):
         """Provide a PHE-based TabPFN regressor as the estimator."""
         # For PHE, we can make tests faster by limiting time and using minimal models
-        max_time = 1 if FAST_TEST_MODE else 5  # Very limited time for fast testing
+        max_time = 15 if FAST_TEST_MODE else 30  # Very limited time for fast testing
```
Comment (noahho) [inline tests/test_post_hoc_ensembles.py]: why increased?

**Review by noahho — CHANGES_REQUESTED** [review]
Nice! This looks great after the small comments are adressed

File: pyproject.toml (hunk 8 above)
Comment (klemens-floege) [reply pyproject.toml:66]: Done

File: tests/test_post_hoc_ensembles.py (hunk 9 above)
Comment (klemens-floege) [reply tests/test_post_hoc_ensembles.py]: If max_time is set too low AutoGluon will fail to fit any models during the fit(). Especially, since we are just building a TabPFN ensemble. This causes errors. I had a few tests failing in some CI hardware so decided it would be more convenient to slightly increase the time limit. After some experimentation got it down to 20: 

`max_time = 10 if FAST_TEST_MODE else 20`

File: tests/test_post_hoc_ensembles.py (hunk 10 above)
Comment (klemens-floege) [reply tests/test_post_hoc_ensembles.py]: See above

File: tests/test_post_hoc_ensembles.py (hunk 9 above)
Comment (klemens-floege) [reply tests/test_post_hoc_ensembles.py]: added comment in test

File: src/tabpfn_extensions/post_hoc_ensembles/sklearn_interface.py [hunk 11]
```diff
@@ -1,141 +1,385 @@
...
+        self,
+        X: pd.DataFrame | np.ndarray,
+        y: pd.Series | np.ndarray,
+        categorical_feature_indices: list[int] | None = None,
+    ) -> AutoTabPFNClassifier:
+        X, y = self._prepare_fit(X, y, categorical_feature_indices)
 
         # Single class case - special handling
-        if len(self.classes_) == 1:
+        if len(unique_labels(y)) == 1:
```
Comment (noahho) [inline src/tabpfn_extensions/post_hoc_ensembles/sklearn_interface.py:369]: Do we need to handle this?

**Review by noahho — APPROVED** [review]
LGTM

File: src/tabpfn_extensions/post_hoc_ensembles/sklearn_interface.py (hunk 11 above)
Comment (klemens-floege) [reply src/tabpfn_extensions/post_hoc_ensembles/sklearn_interface.py:369]: discussed offline
